benchmark/single-source/FindStringNaive.swift

```diff
@@ -0,0 +1,116 @@
+  BenchmarkInfo(
+    name: "FindStringNaive.String",
+    runFunction: runBenchString,
+    tags: t,
+    setUpFunction: {
+      longStringXsThenFoo = String(repeating: "x", count: 10_000) + "foo <-- needle"
+    }),
+  BenchmarkInfo(
+    name: "FindStringNaive.ArrayUTF8",
+    runFunction: runBenchArrayOfUTF8,
+    tags: t,
+    setUpFunction: {
+      longArrayXsThenFoo = Array(repeating: UInt8(ascii: "x"), count: 1_000_000) + Array("foo <-- needle".utf8)
+    }),
+]
+
+func find<S: StringProtocol>(_ string: S,
+                             needle1: Character,
+                             needle2: Character?,
+                             needle3: Character?) -> String.Index? {
+    var index = string.startIndex
+    while index < string.endIndex {
+        if string[index] == needle1 {
+            if let needle2 = needle2 {
+                if find(string[string.index(after: index)...],
+                        needle1: needle2,
+                        needle2: needle3,
+                        needle3: nil) == string.index(after: index) {
+                  return index
+                }
+            } else {
+                return index
+            }
+        }
+        index = string.index(after: index)
+    }
+    return nil
+}
+
+func findOnUTF8Collection<Bytes: Collection>(_ string: Bytes,
+                                             needle1: UInt8,
+                                             needle2: UInt8?,
+                                             needle3: UInt8?) -> Bytes.Index? where Bytes.Element == UInt8 {
+    var index = string.startIndex
+    while index < string.endIndex {
+        if string[index] == needle1 {
+            if let needle2 = needle2 {
+                if findOnUTF8Collection(string[string.index(after: index)...],
+                                        needle1: needle2,
+                                        needle2: needle3,
+                                        needle3: nil) == string.index(after: index) {
+                  return index
+                }
+            } else {
+                return index
+            }
+        }
+        index = string.index(after: index)
+    }
+    return nil
+}
+
+
+@inline(never)
+func runBenchSubstring(iterations: Int) {
+    for _ in 0 ..< iterations {
+        precondition(find(longStringXsThenFoo![...], needle1: "f", needle2: "o", needle3: "o") != nil)
+    }
+}
+
+@inline(never)
+func runBenchString(iterations: Int) {
+    for _ in 0 ..< iterations {
+        precondition(find(longStringXsThenFoo!, needle1: "f", needle2: "o", needle3: "o") != nil)
+    }
+}
+
+@inline(never)
+func runBenchArrayOfUTF8(iterations: Int) {
+    for _ in 0 ..< iterations {
+        precondition(findOnUTF8Collection(longArrayXsThenFoo!,
+                                          needle1: UInt8(ascii: "f"),
+                                          needle2: UInt8(ascii: "o"),
+                                          needle3: UInt8(ascii: "o")) != nil)
+    }
+}
```

Comment:
I think the interesting comparison to make is against a non-recursive benchmark.

```swift
func findNonRec<S: StringProtocol>(
  _ string: S, needle1: Character, needle2: Character?, needle3: Character?
) -> String.Index? {
  var index = string.startIndex
  while index < string.endIndex {
    let nextIndex = string.index(after: index)
    defer { index = nextIndex }

    // Check the first needle
    guard string[index] == needle1 else { continue }

    // Check the second (if applicable)
    guard let needle2 = needle2 else { return index }
    guard nextIndex < string.endIndex else { return nil }
    guard string[nextIndex] == needle2 else { continue }

    // Check the third (if applicable)
    guard let needle3 = needle3 else { return index }
    let nextNextIndex = string.index(after: nextIndex)
    guard nextNextIndex < string.endIndex else { return nil }

    guard string[nextNextIndex] == needle3 else { continue }

    // All out of needles
    return index
  }

  return nil
}

```

Reply:
@milseman you don't need any recursion to trigger the ARC-ache here. The reason the test case is written as is is because I reduced it from a real-world example at https://github.com/adtrevor/ParserBuilder . The code below (please note the commented out lines):

```
func find<S: StringProtocol>(_ string: S,
                             needle1: Character,
                             needle2: Character?,
                             needle3: Character?) -> String.Index? {
  var index = string.startIndex
  while index < string.endIndex {
    var nextIndex = string.index(after: index)
      if string[index] == needle1 {
        // Check subsequent needles recursively (if applicable)
        guard let needle2 = needle2 else { return index }

        //if find(
                //string[nextIndex...], needle1: needle2, needle2: needle3, needle3: nil
               //) == nextIndex {
          return index
        //}
      }
    index = nextIndex
  }
  return nil
}
```

ARCs just as much. The simplest version with the same ARC profile I can come up with is actually

```swift
func find<S: StringProtocol>(_ string: S,
                             needle: Character) -> String.Index? {
  var index = string.startIndex
  while index < string.endIndex {
    var nextIndex = string.index(after: index)
    if string[index] == needle {
      return index
    }
    index = nextIndex
  }
  return nil
}
```

want me to use that instead?